src/calendarworker.cpp

```diff
@@ -746,20 +748,18 @@ CalendarWorker::eventOccurrences(const QList<CalendarData::Range> &ranges) const
 
 QHash<QDate, QStringList>
 CalendarWorker::dailyEventOccurrences(const QList<CalendarData::Range> &ranges,
-                                      const QMultiHash<QString, QDateTime> &allDay,
                                       const QList<CalendarData::EventOccurrence> &occurrences)
 {
     QHash<QDate, QStringList> occurrenceHash;
     foreach (const CalendarData::Range &range, ranges) {
         QDate start = range.first;
         while (start <= range.second) {
             foreach (const CalendarData::EventOccurrence &eo, occurrences) {
+                QTimeZone tz = QTimeZone::systemTimeZone();
                 // On all day events the end time is inclusive, otherwise not
-                if ((eo.startTime.date() < start
-                     && (eo.endTime.date() > start
-                         || (eo.endTime.date() == start && (allDay.contains(eo.eventUid, eo.recurrenceId)
-                                                            || eo.endTime.time() > QTime(0, 0)))))
-                        || (eo.startTime.date() >= start && eo.startTime.date() <= start)) {
+                const bool isAllDay = eo.eventAllDay;
+                if ((isAllDay && eo.startTime.date() <= start && eo.endTime.date() > start) ||
+                   (!isAllDay && eo.startTime.toTimeZone(tz).date() <= start && eo.endTime.toTimeZone(tz).date() >= start)) {
```

Comment:
Maybe one can avoid the conversion to the system time zone here with the following:
```
const QDateTime startDt(start); // Will be later start.startOfDay()
...
eo.startTime <= startDt && eo.endTime >= startDt
```

Reply:
Sorry, I was just speaking about the line with the timezone change. Something like:
```
const QDateTime startDt(start); // Will be later start.startOfDay()
const bool isAllDay = eo.eventAllDay;
if ((isAllDay && eo.startTime.date() <= start && eo.endTime.date() > start) ||
    (!isAllDay && eo.startTime <= startDt && eo.endTime >= startDt)) {
```
Like that you're doing QDateTime comparisons and it's guarantee to work if time zones are different. Instead of using the date which requires to adjust the timezones. By creating startDt from start, you're defining 00:00 of that day in the system timezone. Then, you can compare (for non all day events) this startDt with startTime and endTime, even if they are in another timezone.

Reply:
Seems to understand you.

it won't work that way.
you need two variables.   startDt  and endDt
QDateTime endDt = startDt.addDays(1);

or it is necessary to complicate the condition.  
............(!isAllDay && eo.startTime <= startDt && eo.endTime >= startDt.addDays(1)).....

it doesn't matter to me. It is possible and so and so.   
But the variant with timezone more clearly shows what is happening here and why.